### PR TITLE
[Feature] Jira

### PR DESCRIPTION
**Pull request by Getriax**
**Task description**

Integration with jira api and implementation:

`jira issues [of @user] [status text] [page number]` - view the issues of the user
**alias: jira i** 

**Example commands**

* `jira issues` 
   * Displays all your jira issues - first page - 10 per page
*  `jira issues page 5` 
   * Displays all your jira issues - page 5
*  `jira issues status in progress`
   * Displays all your jira issues with status 'in progress' - first page
*  `jira issues status done page 2` 
   * Displays all your jira issues with status 'done' - page 2
*  `jira issues of @Josh` 
   * Displays all issues of Josh - first page - 10 per page
*  `jira issues of @Alice page 3` 
   * Displays all issues of Alice - page 3
* `jira issues of @Josh status on hold` 
   * Displays all issues of Josh with status 'on hold' - first page
* `jira issues of @Alice status done page 5` 
   * Displays all issues of Alice with status 'done' - page 5'

**Required envs:**
* HUBOT_ATLASSIAN_AUTH_TOKEN
   * token for authentication with atlassian
* HUBOT_ATLASSIAN_USERNAME
   * username/email of atlassian user
* HUBOT_ATLASSIAN_URL
   *  the url for the atlassian

### CODE REVIEW

File: src/jira/jira.api.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  getParams(assignee, statusName, page = 1) {
+    const params = {
+      jql: `assignee="${assignee}"`,
+      fields: query.fields.join(','),
+      maxResults: this.issuesOnPage,
+      startAt: 10 * (page - 1),
+    };
+
+    if (statusName) {
```
Comment (krzyurb) [inline src/jira/jira.api.js]: What is `jql`? Consider delegating lines `74`-`78` into separated function with descriptive name (e.g. `preparedJiraQuerry(statusName)`) that returns final value in `68`

File: src/jira/jira.errors.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const { BotResponseError } = require('../../utils');
+
+const errors = {
+  SERVER: new BotResponseError('Connection with server error', 'It seems like I have a problem connecting to the atlassian server :('),
+  AUTHENTICATION: new BotResponseError('Authentication error', 'I\'m not authenticated, configuration problem, try again later!'),
+  NO_RESULTS: new BotResponseError('No results error', 'Oops... it seems like there is no issues'),
```
Comment (krzyurb) [inline src/jira/jira.errors.js]: Shold be: _ops... it seems like there **are** no issues_

File: src/jira/jira.service.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    return res.message.user.name;
+  }
+
+  extractUserFromMentions(res) {
+    return res.message.mentions.filter(user => user.type === 'user')[0].info.name;
+  }
+
+  extractQueryData(res, match) {
+    const assignee = match[3] ? this.extractUserFromMentions(res) : this.extractSelfUser(res);
+    const status = match[6];
```
Comment (krzyurb) [inline src/jira/jira.service.js]: What if `match[6]` will be undefined?

File: src/jira/index.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+module.exports = (robot) => {
+  robot.respond(/jira .*/i, async (res) => {
+    const { JiraService } = robot.brain;
+    const { logger } = robot;
+    const controller = new JiraController(JiraService);
+
+    const multiplexer = new RespondMultiplexer(logger);
+
+    multiplexer.addResponse(/jira i(ssues)?( of (@((?!page|status)\w+.?)*))?(\s?status (((?!page)\w+.?)*))?(\s?page (\d))?/i, controller.searchForUserIssues);
```
Comment (krzyurb) [inline src/jira/index.js]: You are using unnecessarily so many capture groups. 
For `jira issues page 5` you get:
```
[0] jira issues page 5
[1] ssues
[2] page 5
[3] 5
```

Consider using [non capturing groups](https://www.stefanjudis.com/today-i-learned/non-capturing-groups-in-regular-expressions/) like: `/jira i(?:ssues)?...`

File: src/jira/jira.api.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  getTotalPages(total) {
+    return Math.ceil(total / this.issuesOnPage);
+  }
+
+  getParams(assignee, statusName, page = 1) {
+    const params = {
+      jql: `assignee="${assignee}"`,
+      fields: query.fields.join(','),
+      maxResults: this.issuesOnPage,
+      startAt: 10 * (page - 1),
```
[the remark this answers is not in the document]
Comment (Getriax) [reply src/jira/jira.api.js]: That is right 😅

File: src/jira/jira.errors.js (hunk 2 above)
Comment (Getriax) [reply src/jira/jira.errors.js]: That is correct 😅

File: src/jira/jira.service.js (hunk 3 above)
Comment (Getriax) [reply src/jira/jira.service.js]: If the status is undefined all statuses will be fetched

File: src/jira/index.js (hunk 4 above)
Comment (Getriax) [reply src/jira/index.js]: Fixed!

File: src/jira/jira.api.js (hunk 1 above)
Comment (Getriax) [reply src/jira/jira.api.js]: Great idea.
`jql: this.prepareJiraQuery(assignee, statusName)` should be better? 😄

File: config/jira.config.json [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  },
+  "colors": {
+    "Blocker":"#FF0000",
+    "Critical":"#FF3232",
+    "High":"#FFD532",
+    "Medium":"#0000FF",
+    "Low":"#00FF00",
+    "Very Low":"#E8D6C0",
+    "default":"#F7E4CC",
+    "page":"#010101"
```
[the remark this answers is not in the document]
Comment (Getriax) [reply config/jira.config.json]: Also replaced **spaces** with **_**

**Comment by Getriax** [comment]
@krzyurb fixes done!

File: test/jira.spec.js [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+
+    const message = {
+      mentions: [user, notUser],
+    };
+
+    const actual = jiraService.extractUserFromMentions({ message });
+
+    const expected = user.info.name;
+
+    expect(actual).to.eql(expected);
```
Comment (Davlis) [inline test/jira.spec.js]: You shouldn't mix unit and integration tests. Maybe we could create `unit` and `integration` folders to separate these tests? Or you can change them to be integration tests

File: src/jira/jira.controller.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable class-methods-use-this */
+
```
Comment (Davlis) [inline src/jira/jira.controller.js]: Consider removing this space

File: scripts/jira.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
+// TODO Change Description
```
Comment (Davlis) [inline scripts/jira.js]: Please remove //TODO

File: src/jira/jira.api.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  prepareJiraQuery(assignee, statusName) {
+    let jql = `assignee="${assignee}"`;
+
+    if (statusName) {
+      jql += ` and status="${statusName}"`;
+    }
+
+    jql += this.orderBy(query.orderBy);
```
Comment (Davlis) [inline src/jira/jira.api.js]: `orderBy` is strongly coupled with `prepareJiraQuery`, would it even work if we invoke `orderBy` at beginning? (becase we would have space in jql)

Why do we even need function for `orderBy` when you didn't create function for `statusName` or `assignee`? Maybe we could drop that function and prepare full query in `prepareJiraQuery` function?

File: src/jira/jira.errors.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+const { BotResponseError } = require('../../utils');
+
+const errors = {
+  SERVER: new BotResponseError('Connection with server error', 'It seems like I have a problem connecting to the atlassian server :('),
+  AUTHENTICATION: new BotResponseError('Authentication error', 'I\'m not authenticated, configuration problem, try again later!'),
+  NO_RESULTS: new BotResponseError('No results error', 'Ops... it seems like there are no issues'),
+  NO_ASSIGNEE: new BotResponseError('No assigne specified', 'Jira issues? Of who...? Make sure you are authenticated!'),
```
Comment (Davlis) [inline src/jira/jira.errors.js]: Why we need to make sure if we are authenticated if there is no assignee? Isn't this problem of bad RegExp for catching command?

File: test/jira.spec.js (hunk 7 above)
Comment (Getriax) [reply test/jira.spec.js]: Switched those tests to integration ones, also fixed the case when `priority` is **NULL**

**Comment by Getriax** [comment]
@Davlis fixes done!

**Pull request merged**